feature: 'spack find --info' with alias 'spack info --installed'

tldahlgren opened the pull request:
People have requested a command to get details on how a package is actually built to include, more recently, compiler flags.  Unless I'm mistaken, that level of detail is not available in any Spack commands so this PR leverages `spack info` to provide it.

(Update: `spack find --json` *does* provide all of the information here and a bit more in machine-readable form.  `spack find` also has options to get dependencies (`-d`), variants (`-v`), compiler flags (`-f`), install path (`-p`).  I still think the  output from this PR is more user friendly and complete -- except when `--json` is used with `spack find`.)

Help output:
```
$ spack info --help
usage: spack info [-hi] ...

get detailed information on a particular package

positional arguments:
  pkg_spec         Package name or, with --installed, spec

optional arguments:
  -h, --help       show this help message and exit
  -i, --installed  Show installation details: same as `spack find --info`
```

Output from a Spack instance with one version of _hypre_ 2.23.0 installed:
```
$ spack find --info hypre@2.23.0      # OR  $ spack info --installed hypre@2.23.0
AutotoolsPackage:   hypre@2.23.0

Hashes:
    DAG  = odn3zibu5sjt2xumsin6b4quduykiklt
    full = jtv3ffez3sjcrwgrvrtdcxjndrczdhz6

Installed Explicitly:   Thursday 21 October 2021
    architecture = linux-rhel7-skylake_avx512
    prefix       = /usr/WS1/dahlgren/spack/opt/spack/linux-rhel7-skylake_avx512/gcc-8.3.1/hypre-2.23.0-odn3zibu5sjt2xumsin6b4quduykiklt

Compiler:   gcc@8.3.1
    cflags = []
    cppflags = []
    cxxflags = []
    fflags = []
    ldflags = []
    ldlibs = []

Variant Settings:
    ~complex
    ~cuda
    cuda_arch=none
    ~debug
    +fortran
    ~int64
    ~internal-superlu
    ~mixedint
    +mpi
    ~openmp
    +shared
    ~superlu-dist
    ~unified-memory

Recorded Dependencies:
    ('build', 'link') dependencies on openblas
    ('build', 'link') dependencies on openmpi
```

Output when multiple versions of a package are installed in the instance:
```
$ spack find --info zlib
==> Error: Information is provided for a single package only

$ spack info -i zlib
==> Error: zlib matches multiple packages.
  Matching packages:
    s4yfeca zlib@1.2.11%clang@11.0.1 arch=linux-rhel7-cascadelake
    c7zv5ft zlib@1.2.11%gcc@8.3.1 arch=linux-rhel7-broadwell
    ash26ww zlib@1.2.11%gcc@8.3.1 arch=linux-rhel7-skylake_avx512
    kzsxhof zlib@1.2.11%gcc@8.3.1 arch=linux-rhel7-skylake_avx512
  Use a more specific spec.

$ spack find --info /s4yfeca     # OR  $ spack info -i /s4yfeca
Package:   zlib@1.2.11

Hashes:
    DAG  = s4yfecahwfhlopnflf3unufea6hh2efg
    full = n2uzqshm22zbhvho3lw36zrtziglracf

Installed as Dependency:   Wednesday 01 September 2021
    architecture = linux-rhel7-cascadelake
    prefix       = /usr/WS1/dahlgren/spack/opt/spack/linux-rhel7-cascadelake/clang-11.0.1/zlib-1.2.11-s4yfecahwfhlopnflf3unufea6hh2efg

Compiler:   clang@11.0.1
    cflags = []
    cppflags = []
    cxxflags = []
    fflags = []
    ldflags = []
    ldlibs = []

Variant Settings:
    +optimize
    +pic
    +shared

Recorded Dependents:
    libxml2 ('build', 'link') dependencies on zlib
```

Output for an externally installed package:
```
$ spack find --info /er475uw    # OR $ spack info -i /er475uw
AutotoolsPackage:   xz@5.2.2

Hashes:
    DAG  = er475uwcmw5la52ymtpaskn43zua5x37
    full = ydmq7okvsa4c3z6hffqxnki42pnymwkt

Installed as Dependency:   Friday 30 July 2021
    architecture = linux-rhel7-broadwell
    prefix       = /usr

External:
    path             = /usr
    module           = None
    extra_attributes = {}

Compiler:   gcc@8.3.1
    cflags = []
    cppflags = []
    cxxflags = []
    fflags = []
    ldflags = []
    ldlibs = []

Variant Settings:
    libs=shared,static
    ~pic

Recorded Dependents:
    libxml2 ('build', 'link') dependencies on xz
    gettext ('build', 'link', 'run') dependencies on xz
```


TODO:
- [x] Is it better to use the ASP solver or database to get to the information?  (A: database is faster and more complete)
- [x] Format the output, whether ASP or database, in a human-readable form

tldahlgren committed:
feature/spack info --installed

tgamblin commented:
@adamjstewart thoughts?

adamjstewart commented:
I agree that we need a command to display this kind of information. But I think this might belong in `spack find`. In my mind, `spack info` gives info about a build recipe, while `spack find` finds details of installed packages.

tldahlgren commented:
@tgamblin @adamjstewart 

The intention here was to have a command that provides human-readable output for a *single* installation, not multiple as I expect to be supported by `spack find`.

I will hold off addressing check issues until an explicit decision to retain -- so go forward with -- or close this PR is made.

tldahlgren committed:
Support 'spack find --info' with 'spack info --installed' as alias

tldahlgren committed:
Fix readthedocs and style errors

tldahlgren committed:
Added unit tests (to increase coverage)

tldahlgren commented:
> @tgamblin @adamjstewart
> 
> The intention here was to have a command that provides human-readable output for a _single_ installation, not multiple as I expect to be supported by `spack find`.
> 
> I will hold off addressing check issues until an explicit decision to retain -- so go forward with -- or close this PR is made.

Looks like the note I thought I added over a week ago didn't post.  

I took a page out of Todd's book from a recent PR and changed this PR to be available to `spack find` *and* "aliased" in `spack info`.

tldahlgren commented:
@adamjstewart Is this alias-based approach more acceptable?

tldahlgren commented:
Forcing re-running of checks since the `package-audits` never reported its status.

tldahlgren commented:
Closing this PR due to the approach being "unpopular".
